### DISCUSSION TITLE
New TuyaMCU Protokoll? V4.0.2? Help

### DISCUSSION THREAD

**Discussion by gummiklaus** [arendst/Tasmota#21095] [Support request]
Hello,

first of all sorry I can hardly speak English and this is translated with DeepL.
So I have a SmartEnergySensor SPM01-D2TW here.

[https://blakadder.com/zemismart-spm01/](url)

I have replaced the CB2S with an ESP02S with Tasmota13.4.
TuyaMCU(54) selected. GPIO1 set to TuyaTX and GPIO3 set to TuyaRX.
With the help of Tasmota-Tuya-Helper, tried to identify the data points. Unfortunately I fail here mercilessly.

There seems to be a new protocol. After switching on, it tells me in the console:
MCU Product ID: {"p": "aqktrdhlsw8q90dr", "v": "4.0.2", "m":0, "n":0, "low":1}

Voltage and current now appear to be at DpId 102 and 103. The mains frequency seems to come from data point 101, but the power is probably at 104 but not in deciWatt. There is a value under 15 that seems to be the cos.

There is also no continuous transmission, I can update the values with TuyaSend0. However, this also produces an "RX unknown command".

Is there already something known or am I wrong here. 
It would be nice if this could be made to work.

Thanks Klaus

Is Tasmota Start:
> 10:01:51.062 HTP: Webserver aktiv bei tasmota-xxxxxx-7817 mit IP-Adresse 192.168.xxx.xxx
10:01:52.041 RSL: INFO1 = {"Info1":{"Module":"TUYA SPM01","Version":"13.4.0(tasmota)","FallbackTopic":"cmnd/DVES_445E89_fb/","GroupTopic":"cmnd/tasmotas/"}}
10:01:52.043 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-xxxxxx-7817","IPAddress":"192.168.xxx.xxx"}}
10:01:52.046 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":4}}
10:01:52.048 RSL: RESULT = {"POWER":"OFF"}
10:01:52.049 RSL: POWER = OFF
10:01:52.125 TYA: MCU Product ID: {"p":"aqktrdhlsw8q90dr","v":"4.0.2","m":0,"n":0,"low":1}
10:01:53.016 QPC: Reset
10:01:56.912 RSL: STATE = {"Time":"2024-04-04T10:01:56","Uptime":"0T00:00:10","UptimeSec":10,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":31,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"xxxxxxxx","BSSId":"xx:xx:xx:xx:xx:xx","Channel":11,"Mode":"11n","RSSI":86,"Signal":-57,"LinkCount":1,"Downtime":"0T00:00:04"}}
10:02:18.014 CMD: weblog4
10:02:18.019 RSL: RESULT = {"WebLog":2}
10:02:45.203 CMD: weblog 4
10:02:45.207 RSL: RESULT = {"WebLog":4}
10:02:45.737 CFG: in Flash gespeichert an F6, Anzahl 14, Bytes 4096
10:02:45.900 TYA: Send "55aa00000000ff"
10:02:45.917 TYA: Send "55aa001c0008011804040a022d0481"
10:02:45.919 DMP: 55 AA 03 00 00 01 01 04
10:02:45.921 {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
10:02:45.922 TYA: Heartbeat
10:02:46.873 WIF: Sending Gratuitous ARP
10:02:47.066 DMP: 55 AA 03 2B 00 00 2D
10:02:47.068 {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
10:02:47.077 TYA: Send "55aa002b0001032e"
10:02:48.164 DMP: 55 AA 03 2B 00 00 2D
10:02:48.166 {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
10:02:48.176 TYA: Send "55aa002b0001032e"
10:02:50.792 WIF: Prüfe Verbindung…
10:02:52.172 DMP: 55 AA 03 2B 00 00 2D
10:02:52.174 {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
10:02:52.184 TYA: Send "55aa002b0001032e"
10:02:53.271 DMP: 55 AA 03 2B 00 00 2D
10:02:53.272 {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
10:02:53.282 TYA: Send "55aa002b0001032e"
10:02:56.911 TYA: Send "55aa00000000ff"
10:02:56.920 DMP: 55 AA 03 00 00 01 01 04
10:02:56.922 {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
10:02:56.923 TYA: Heartbeat
10:02:57.252 DMP: 55 AA 03 2B 00 00 2D
10:02:57.254 {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
10:02:57.264 TYA: Send "55aa002b0001032e"

Is send TuyaSend0

> 16:33:10.397 CMD: TuyaSend0
16:33:10.399 SRC: WebConsole from 192.168.xxx.xxx
16:33:10.402 CMD: Grp 0, Cmd 'TUYASEND', Idx 0, Len 0, Pld -99, Data ''
16:33:10.406 TYA: Read MCU state
16:33:10.415 TYA: Send "55aa0008000007"
16:33:10.418 RSL: RESULT = {"TuyaSend":"Done"}
16:33:10.512 TYA: RX value 0 from dpId 2 
16:33:10.515 DMP: 55 AA 03 07 00 0C 06 00 00 08 09 33 00 05 06 00 01 28 93
16:33:10.518 RSL: RESULT = {"TuyaReceived":{"Data":"55AA0307000C06000008093300050600012893","Cmnd":7,"CmndData":"060000080933000506000128","DpType0Id6":"0x0933000506000128","6":{"DpId":6,"DpIdType":0,"DpIdData":"0933000506000128"}}}
16:33:10.522 TYA: fnId=0 is set for dpId=6
16:33:10.524 DMP: 55 AA 03 07 00 08 0F 02 00 04 00 00 00 61 87
16:33:10.527 RSL: RESULT = {"TuyaReceived":{"Data":"55AA030700080F0200040000006187","Cmnd":7,"CmndData":"0F02000400000061","DpType2Id15":97,"15":{"DpId":15,"DpIdType":2,"DpIdData":"00000061"}}}
16:33:10.530 TYA: fnId=0 is set for dpId=15
16:33:10.532 TYA: RX value 97 from dpId 15 
16:33:10.534 DMP: 55 AA 03 07 00 05 09 05 00 01 00 1D
16:33:10.537 RSL: RESULT = {"TuyaReceived":{"Data":"55AA0307000509050001001D","Cmnd":7,"CmndData":"0905000100","DpType5Id9":"0x00","9":{"DpId":9,"DpIdType":5,"DpIdData":"00"}}}
16:33:10.540 TYA: fnId=0 is set for dpId=9
16:33:10.543 DMP: 55 AA 03 07 00 05 10 01 00 01 01 21
16:33:10.546 RSL: RESULT = {"TuyaReceived":{"Data":"55AA03070005100100010121","Cmnd":7,"CmndData":"1001000101","DpType1Id16":1,"16":{"DpId":16,"DpIdType":1,"DpIdData":"01"}}}
16:33:10.549 TYA: fnId=0 is set for dpId=16
16:33:10.551 DMP: 55 AA 03 07 00 0C 11 00 00 08 07 01 00 0D 08 00 00 00 4B
16:33:10.555 RSL: RESULT = {"TuyaReceived":{"Data":"55AA0307000C110000080701000D080000004B","Cmnd":7,"CmndData":"110000080701000D08000000","DpType0Id17":"0x0701000D08000000","17":{"DpId":17,"DpIdType":0,"DpIdData":"0701000D08000000"}}}
16:33:10.558 TYA: fnId=0 is set for dpId=17
16:33:10.561 DMP: 55 AA 03 07 00 18 12 00 00 14 01 01 00 3C 03 01 00 FD 04 00 00 B4 07 01 00 00 09 00 00 00 4F
16:33:10.565 RSL: RESULT = {"TuyaReceived":{"Data":"55AA03070018120000140101003C030100FD040000B407010000090000004F","Cmnd":7,"CmndData":"120000140101003C030100FD040000B40701000009000000","DpType0Id18":"0x0101003C030100FD040000B40701000009000000","18":{"DpId":18,"DpIdType":0,"DpIdData":"0101003C030100FD040000B40701000009000000"}}}
16:33:10.568 TYA: fnId=0 is set for dpId=18
16:33:10.575 DMP: 55 AA 03 07 00 08 65 02 00 04 00 00 13 94 23
16:33:10.579 RSL: RESULT = {"TuyaReceived":{"Data":"55AA03070008650200040000139423","Cmnd":7,"CmndData":"6502000400001394","DpType2Id101":5012,"101":{"DpId":101,"DpIdType":2,"DpIdData":"00001394"}}}
16:33:10.582 TYA: fnId=0 is set for dpId=101
16:33:10.584 TYA: RX value 5012 from dpId 101 
16:33:10.591 DMP: 55 AA 03 07 00 08 66 02 00 04 00 00 09 33 B9
16:33:10.595 RSL: RESULT = {"TuyaReceived":{"Data":"55AA030700086602000400000933B9","Cmnd":7,"CmndData":"6602000400000933","DpType2Id102":2355,"102":{"DpId":102,"DpIdType":2,"DpIdData":"00000933"}}}
16:33:10.598 TYA: fnId=33 is set for dpId=102
16:33:10.600 TYA: RX value 2355 from dpId 102 
16:33:10.602 TYA: Rx ID=102 Voltage=2355
16:33:10.606 DMP: 55 AA 03 07 00 08 67 02 00 04 00 00 05 06 89
16:33:10.610 RSL: RESULT = {"TuyaReceived":{"Data":"55AA03070008670200040000050689","Cmnd":7,"CmndData":"6702000400000506","DpType2Id103":1286,"103":{"DpId":103,"DpIdType":2,"DpIdData":"00000506"}}}
16:33:10.613 TYA: fnId=32 is set for dpId=103
16:33:10.615 TYA: RX value 1286 from dpId 103 
16:33:10.617 TYA: Rx ID=103 Current=1286
16:33:10.622 DMP: 55 AA 03 07 00 08 68 02 00 04 00 00 01 28 A8
16:33:10.626 RSL: RESULT = {"TuyaReceived":{"Data":"55AA030700086802000400000128A8","Cmnd":7,"CmndData":"6802000400000128","DpType2Id104":296,"104":{"DpId":104,"DpIdType":2,"DpIdData":"00000128"}}}
16:33:10.628 TYA: fnId=31 is set for dpId=104
16:33:10.630 TYA: RX value 296 from dpId 104 
16:33:10.632 TYA: Rx ID=104 Active_Power=296
16:33:10.634 DMP: 55 AA 03 2D 00 00 2F
16:33:10.637 RSL: RESULT = {"TuyaReceived":{"Data":"55AA032D00002F","Cmnd":45}}
16:33:10.639 TYA: RX unknown command
16:33:10.648 DMP: 55 AA 03 2B 00 00 2D
16:33:10.651 RSL: RESULT = {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
16:33:10.661 TYA: Send "55aa002b0001032e"
16:33:14.267 TYA: Send "55aa00000000ff"
16:33:14.276 DMP: 55 AA 03 00 00 01 01 04
16:33:14.279 RSL: RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
16:33:14.282 TYA: Heartbeat
16:33:14.620 DMP: 55 AA 03 2B 00 00 2D
16:33:14.623 RSL: RESULT = {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
16:33:14.633 TYA: Send "55aa002b0001032e"
16:33:15.707 DMP: 55 AA 03 2B 00 00 2D
16:33:15.710 RSL: RESULT = {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
16:33:15.721 TYA: Send "55aa002b0001032e"
16:33:19.706 DMP: 55 AA 03 2B 00 00 2D
16:33:19.709 RSL: RESULT = {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
16:33:19.720 TYA: Send "55aa002b0001032e"
16:33:20.819 DMP: 55 AA 03 2B 00 00 2D

**Comment by RaW79** [comment]
Hi Klaus, have identified any further solutions in this Case? Like this Energy Meter very much and feel confident in replacing to ESP02s but regarding the Tuya MCU i have nö experience at all. So in Case you can inform about News IT would be a great Help for me. Thx in advance

**Reply by gummiklaus** [reply]
Hi, unfortunately no news :-(
ESP02 is in with Tasmota. Unfortunately, Tasmota does not yet know the new protocol.
So far I could only find out that the TuyaMCU requires an active connection. It probably expects the registration.
If I have some air I want to connect the CB2S parallel to the ESP and set the ESP serial read only to see what is going on there. 
If you can do something with the data, I will be happy to send it to you

Translated with DeepL.com (free version)

**Reply by RaW79** [reply]
Thanks for your Feedback. Maybe next week my devices will be delivered and then I'll follows you in Testing. But due to nearly no expirience of mine in interfacing Tuya MCUs expecting Not to much Outcome. 

Currently your information is beside this articel the only information i could find: https://blakadder.com/zemismart-spm01/

With regard to your Info this Sounds a bit too easy for me in the articel where no problem ist Seen by replacing the WiFi Version with an ESP02s and blakadder is not a nobody in the world of Tasmota. Therefore I'm a little bit surprised and don't know what to expect. WE will see and let you know of course

**Comment by RaW79** [comment]
Hallo Klaus,

bin jetzt mit dem Hardwareumbau durch und ein erster Versuch sieht für mich noch schlechter als bei Dir aus, da ich noch nichtmal die Spannung identifizieren kann bei den DpId´s. Was mir auffällt, die andere Version bei der Tuya MCU v3.0.2 gegenüber v4.0.2 bei dir:

Ansonsten fang ich gerade erst an und dies mein Log:

18:09:39.142 TYA: MCU Product ID: {"p":"brnwqs0ifofy3rzu","v":"3.0.2","m":0,"n":0,"low":1}
18:09:39.150 TYA: Send "55aa0002000001"
18:09:39.158 DMP: 55 AA 03 02 00 00 04
18:09:39.159 {"TuyaReceived":{"Data":"55AA0302000004","Cmnd":2}}
18:09:39.161 TYA: RX MCU configuration Mode=0
18:09:39.162 TYA: Read MCU state
18:09:39.170 TYA: Send "55aa0008000007"
18:09:39.187 DMP: 55 AA 03 07 00 08 01 02 00 04 00 00 00 00 18
18:09:39.189 {"TuyaReceived":{"Data":"55AA03070008010200040000000018","Cmnd":7 "CmndData":"0102000400000000","DpType2Id1":0,"1":{"DpId":1,"DpIdType":2,"DpIdData":"00000000"}}}

18:09:39.191 TYA: fnId=11 is set for dpId=1
18:09:39.192 TYA: RX value 0 from dpId 1 
18:09:39.202 DMP: 55 AA 03 07 00 08 02 02 00 04 00 00 00 00 19
18:09:39.204 {"TuyaReceived":{"Data":"55AA03070008020200040000000019","Cmnd":7,"CmndData":"0202000400000000","DpType2Id2":0,"2":{"DpId":2,"DpIdType":2,"DpIdData":"00000000"}}}

18:09:39.206 TYA: fnId=0 is set for dpId=2
18:09:39.208 TYA: RX value 0 from dpId 2 
18:09:39.222 DMP: 55 AA 03 07 00 0C 06 00 00 08 09 23 00 00 00 00 00 00 4F
18:09:39.224 {"TuyaReceived":{"Data":"55AA0307000C0600000809230000000000004F","Cmnd":7,"CmndData":"060000080923000000000000","DpType0Id6":"0x0923000000000000","6":{"DpId":6,"DpIdType":0,"DpIdData":"0923000000000000"}}}

18:09:39.227 TYA: fnId=0 is set for dpId=6
18:09:39.235 DMP: 55 AA 03 07 00 05 09 05 00 01 00 1D
18:09:39.238 {"TuyaReceived":{"Data":"55AA0307000509050001001D","Cmnd":7,"CmndData":"0905000100","DpType5Id9":"0x00","9":{"DpId":9,"DpIdType":5,"DpIdData":"00"}}}

18:09:39.241 TYA: fnId=0 is set for dpId=9
18:09:39.247 DMP: 55 AA 03 07 00 05 0C 01 00 01 00 1C
18:09:39.250 {"TuyaReceived":{"Data":"55AA030700050C010001001C","Cmnd":7,"CmndData":"0C01000100","DpType1Id12":0,"12":{"DpId":12,"DpIdType":1,"DpIdData":"00"}}}

18:09:39.252 TYA: fnId=0 is set for dpId=12
18:09:39.263 DMP: 55 AA 03 07 00 08 0D 02 00 04 00 00 00 00 24
18:09:39.265 {"TuyaReceived":{"Data":"55AA030700080D0200040000000024","Cmnd":7,"CmndData":"0D02000400000000","DpType2Id13":0,"13":{"DpId":13,"DpIdType":2,"DpIdData":"00000000"}}}

18:09:39.268 TYA: fnId=0 is set for dpId=13
18:09:39.270 TYA: RX value 0 from dpId 13 
18:09:39.283 DMP: 55 AA 03 07 00 0C 11 00 00 08 07 01 00 0D 08 00 00 00 4B
18:09:39.286 {"TuyaReceived":{"Data":"55AA0307000C110000080701000D080000004B","Cmnd":7,"CmndData":"110000080701000D08000000","DpType0Id17":"0x0701000D08000000","17":{"DpId":17,"DpIdType":0,"DpIdData":"0701000D08000000"}}}

18:09:39.289 TYA: fnId=0 is set for dpId=17
18:09:39.319 DMP: 55 AA 03 07 00 1C 12 00 00 18 01 01 00 3C 03 01 00 FD 04 00 00 B4 07 01 00 00 08 00 00 1E 09 00 00 00 7D
18:09:39.323 {"TuyaReceived":{"Data":"55AA0307001C120000180101003C030100FD040000B4070100000800001E090000007D","Cmnd":7,"CmndData":"120000180101003C030100FD040000B4070100000800001E09000000","DpType0Id18":"0x0101003C030100FD040000B4070100000800001E09000000","18":{"DpId":18,"DpIdType":0,"DpIdData":"0101003C030100FD040000B4070100000800001E09000000"}}}

18:09:39.326 TYA: fnId=0 is set for dpId=18
18:09:39.328 DMP: 55 AA 03 2D 00 00 2F
18:09:39.330 {"TuyaReceived":{"Data":"55AA032D00002F","Cmnd":45}}
18:09:39.332 TYA: RX unknown command
18:09:39.731 DMP: 55 AA 03 2B 00 00 2D
18:09:39.733 {"TuyaReceived":{"Data":"55AA032B00002D","Cmnd":43}}
18:09:39.743 TYA: Send "55aa002b0001032e"

**Reply by gummiklaus** [reply]
Hallo RaW79,

schön das Du auf deutsch schreibst, ich breche mir da mit Englisch die Pfoten ;-)

Also das ist ja Spannend. Welchen Typ hast Du genau?
Ich habe ein ZgMiSmart SPM01 (SPM01-D2Tw)

In Tasmota hab ich TuyaMCU (54) eingestellt mit 
GPIO1 -> Tuya TX
GPIO3 -> Tuya RX

Hast Du den vorher mit der Tuya App gekoppelt?
Bei mir Blinkt die grüne LED, wenn ich das richtig deute bedeutet das die MCU auf die Kopplung wartet.

DpId's hab ich mit hilfe des Tuya2Tasmota Script welches als Lesezeichen eingebunden wird gefunden.
Leider finde ich auf die schnelle den Link nicht wieder zu dieser seite.

**Reply by RaW79** [reply]
Hallo Klaus,

Dein Auszug aus der Konsole auf Deutsch hat mich auf die Fährte gebracht und solange wir 1:1 uns austauschen.... so gut ist mein Englisch nun auch nicht ;)

Ich hab auch einen SPM01-D2TW  und hin nie mit CB2S oder Tuya App betrieben. Ja meine LED blinkt auch.
Da das Wifi aber der ESP/CB2S regelt bin ich mir da unsicher ob das wirklich ein Grund dafür ist.

Den Tuya Helper hattest Du Oben erwähnt und ich hab ihn probiert ohne wirklich dadurch mehr zu erzielen. 
![grafik](https://github.com/arendst/Tasmota/assets/16465702/b478fa38-a4ac-4afb-a739-d528bd23ab09)


Es werden 8 Datenpunkte identifiziert und 3 davon (dpID: 6,17 & 18) mit irgendwelchem Inhalt.

Alleinig dpID 6 mit 0923000000000000 könnte ich mir als Hexadezimal Wert für 233,9V vorstellen.

Die beiden anderen dpIds (17 & 18) bleiben auch ohne angeschlossene Last so wie sie sind und haben dann wohl nichts mit Strom oder Leistung zu tun.

Hab mal hier auf Github versucht mir die Datenpunkte der Zigbee Variante anzusehen und daraus sehe ich Strom, Spannung und Leistung  kann mich aber täuschen. 

Hast Du eine Idee was der Switch (Boolean) für eine Funktion haben könnte? Hab mal fröhlich per TuyaHelper An/Aus geschaltet aber keine Effekte bemerkt.

**Reply by gummiklaus** [reply]
Hmm weicht von meinem ab.

![tuya2tasmota](https://github.com/arendst/Tasmota/assets/58884809/8ca5b048-4cdf-46c3-bdae-8d8991d558f8)

Acha ja die Werte werden nur einmal aktuallsiert.

Ich wollte wenn ich mal etwas Zeit habe den CB2S parallel anschießen und tasmota nur lesend stellen. Bin mir aber nicht sicher ob ich 
damit beide Seiten (Sender mcu und Sender CB2S) sehen kann...

Switch Bolen glaub ich ist nur die erweiterung falls da auch ein Relai drin sein sollte, so würde ich das sehen

**Reply by RaW79** [reply]
Oja das ist ein Unterschied.

Kommen die Daten bei Dir automatisch von der MCU?  Bei mir kommt nichts dieser Daten nur auf Anforderung tuyasend0

Hab den zweiten Zähler nun mal mit CB2S in Betrieb genommen und bei Tuya eingebunden. Was zu erwarten war, 3 Übertragungswerte sind sichtbar. Allerdings ist Vorsicht bei der Einheit der Wirkleistung geboten (60W Glühbirne):
![grafik](https://github.com/arendst/Tasmota/assets/16465702/6777ee54-561a-4bcf-abee-07c81813cca7)

Klarer Unterschied die LED blinkt nicht.

Dann mal den Button am Gerät getestet und er kommt definitv über die MCU und wird mit einem Wifireset auch verarbeitet.

05:47:11.813 {"TuyaReceived":{"Data":"55AA0304000006","Cmnd":4}}
05:47:11.814 TYA: RX WiFi Reset
05:47:11.816 SRC: Button
05:47:11.819 CMD: Grp 0, Cmd 'WIFICONFIG', Idx 1, Len 1, Pld 2, Data '2'
05:47:11.822 RSL: RESULT = {"WifiConfig":{"2":"WifiManager"}}

**Reply by gummiklaus** [reply]
Hi, bei mir kommen auch nur Daten wenn ich die Abfrage mit TuyaSend0.

Hab mich leider bisher noch nicht so reinhängen können.
Gestern bin ich noch auf so ein paar Sachen im Nett gestoßen, das wollte ich mal die nächsten Tage
probieren:
https://github.com/openshwprojects/TuyaMCUAnalyzer

Sieht vielsprechend aus, vielleicht bekommt man so etwas mehr aus der Kommunikation zwischen TuyaMCU-CB2S-TuyaAPP raus.

Dazu noch ein Dok von Tuya leider auf Englisch und ich Quale mich damit rum :-(
https://images.tuyacn.com/smart/aircondition/Guide-to-Interworking-with-the-Tuya-MCU.pdf

Sag mal bist Du Dieter P aus dem IOBroker Forum?
https://forum.iobroker.net/topic/74492/tuyamcu-per-tasmota-daten-identifizieren

Hab auch IOBroker bei mir am laufen *g*

**Reply by RaW79** [reply]
Thx, konnte leider auch noch nicht weiterforschen.

Aber Dein erster Link geht genau in meine Denkrichtung.  Die aktuellen Daten zu verstehen wie im IOBrokerForum ist ein Thema (ja das bin ich :)  aber das nächste Thema ist aus meiner Sicht, wie bekommt man die MCU dazu ständig die Daten zu senden etc. Vermute aktuell auch noch irgendwie das es nicht alle bzw. sogar falsche/unbrauchbare Daten sind da sich erkennbar nur 1 dpIx von 3 ändert. Da sich eigentlich immer die Spannung minimal ändert etc. erwarte ich auch Schwankungen bei der Leistung und Strom und nicht wie jetzt 2 fixe Werte. Irgendwas passt da nicht zusammen.

Zurück zum verstehen der MCU etc. Für mich hat da die OpenBeken-Community sehr großes Wissen und Erfahrung. Ich klopfe da mal an und frage ob etwas bzgl. CB2S empfehlen kann ggf. lässt sich sogar mittels der dort entwickelten Tools der CB2S auslesen und so etwas herausfinden.

Bleibe dran VG

**Reply by RaW79** [reply]
Hi

hab mal etwas probiert mit dem TuyMCU Analyzer. Da ich noch keinen getrennten Betrieb der MCU (ohne ESP oder CB2S) hinbekommen hab war es mäßig erfolgreich. Bin da aber auch aktuell wenig optimistisch da das Tool letztlich nicht viel anders aggiert wie dieses Browser-AddIn was Tasmota untersucht.
Problem von allen diesen Tools ist das Raw Data Format und dann noch noch Register wo mehrere Messwerte in einem Register verwurschtelt sind. Genau wie hier in der Anleitung zum MCU Analyzer beschrieben.

Worbei hilft das Programm nicht? Grundsätzlich gibt es eine Situation, in der dieses Programm nicht helfen kann. Einige der fortschrittlicheren Tuya-Stromzähler (höherwertige, bereits DIN-Schienen, in der Preisklasse von 100 PLN oder mehr) verwenden TuyaMCU, um Messungen zu melden, melden aber oft eine Gruppe von Messungen in einer dpID. Dann ist eine solche dpID vom Typ Raw (nicht mehr Value) und enthält mehrere Bytes, in denen (je nach Gerät) Daten wie der aktuelle Wert von Spannung, Strom, Leistung, manchmal auch Frequenz gespeichert sind. Ein solches Rohpaket hat ein geräteabhängiges Format und es ist schwierig, es universeller zu dekodieren. Günstigere Produkte mit Energiemessung wiederum verwenden keine TuyaMCU, sondern BL0937 oder BL0942 werden dort direkt mit dem WLAN-Modul verbunden.

Ich verfolge nun weiter den Ansatz den RAw Data Inhalt irgendwie rauszutüfteln (siehe IOBroker Forum) gelingt das, versuche ich zu verstehen wie man die MCU zur automatischen Bereitstellung dieser Daten bringt.

**Comment by g6094199** [comment]
I'm in the same boat. in parallel i opened a topic on the [OpenBeken board](https://www.elektroda.com/rtvforum/viewtopic.php?p=21091280) and have already an offer to work on support for this unit. so there is a chance to have at least one working firmare for it.

i have an SP02 which is the 3 phase model of yours. there are a few more dplds for each phase there. 3 (voltage, current and power) for each phase + frequency. the rest is still under investigation.

![331714166-c1a66665-3bde-4f02-bf9f-d2efcc3ead76](https://github.com/arendst/Tasmota/assets/37833326/2bd9da93-b96b-440c-b105-0fe3919dc42b)




Ich schreib lieber in Englisch, da das hier eben mehr Leute verstehen (und es auch besser von Suchmaschinen gefunden wird.), aber im Zweifel koennen wir auch parallel fahren. Aber es sollte moeglichst in englisch bleiben, so nehmen potenziell mehre Leute and er Diskussion teil.

**Reply by RaW79** [reply]
Welcome in the boat :)
For openbeken it´s the same for me please see: https://www.elektroda.com/rtvforum/topic4052633.html#21080800

![4565591700_1715767766](https://github.com/arendst/Tasmota/assets/16465702/b439585a-1478-4039-9e64-243815c79d93)

The values of voltage, current and active power are fine for me and it works. But I do not understand how to get the energy counter values (enery total) to start working which are expected to be at dpIx 1. Only zero as integer is provided so far. 

Guessing that the MCU has to receive a command to start counting but could not find any helpful information so far.

Hope we can get further together.

**Comment by g6094199** [comment]
SPM02 is working fine with OpenBeken, so we will get the SPM01 running for sure.

**Comment by RaW79** [comment]
Any news?

I´m having 2 issues transfering my test outcome from openBeken to Tasmota:

1)  how to send/set the status for the Wifi connection & how to send repeating state queries in Tasmota?

`tuyaMcu_defWiFiState 4`
`addRepeatingEvent 5 -1 tuyaMcu_sendQueryState`

2)  DpId6 is RaW Data Point containing power, current & voltage. How to dissamble and align 1 DpId in Tasmota to 3 channels?
OpenBeken uses a predefined function for this

`// Measurements - Dpid 6 "phase_a" - channel RAW_TAC2121C_VCP -> 5,6,7`
`// TAC2121C VoltageCurrentPower Packet`
`// This will automatically set voltage, power and current`
`linkTuyaMCUOutputToChannel 6 RAW_TAC2121C_VCP`
`setChannelType 5 Voltage_div10`
`setChannelLabel 5 "Voltage"`
`setChannelType 6 Power`
`setChannelLabel 6 "Power"`
`setChannelType 7 Current_div1000`
`setChannelLabel 7 "Current"`

**Comment by jocelioalmeidaset** [comment]
Hello, I also have a three-phase SPM02. Follow the previous messages and I was able to understand part of the following data. What matters is the DpIdData value, which is in hexadecimal. In the case of the SPM02, which is three-phase, the voltage, current and power data are in a single field. For example, 089A0000AB000012. The first 4 bits are the voltage. If you convert 089A from hexadecimal to decimal, you have 2202. Dividing by 10, you have 220.2 V. I was unable to decipher the other bits, but this is the logic.

I have Tasmota firmware with ESP-02S.

**Reply by jocelioalmeidaset** [reply]
successfully obtained voltage and current and power
there are 4 bits for voltage x 10
there are 6 bits for current x 1000
there are 6 bits for power (integer value)

in the data below
02:09:20.919 MQT: tele/tasmota_41AF7F/RESULT = {"TuyaReceived":{"Data":"55AA0307000C06000008087A00024A00004031","Cmnd":7,"CmndData":"06000008087A0002 4A000040","DpType0Id6":"0x087A00024A000040","6":{"DpId":6,"DpIdType":0,"DpIdData":"087A00024A000040"}}}
02:09:20.922 TYA: fnId=0 is set for dpId=6
02:09:20.934 DMP: 55 AA 03 07 00 0C 07 00 00 08 08 78 00 01 23 00 00 1F E7
02:09:20.939 MQT: tele/tasmota_41AF7F/RESULT = {"TuyaReceived":{"Data":"55AA0307000C07000008087800012300001FE7","Cmnd":7,"CmndData":"0700000808780001 2300001F","DpType0Id7":"0x087800012300001F","7":{"DpId":7,"DpIdType":0,"DpIdData":"087800012300001F"}}}
02:09:20.941 TYA: fnId=0 is set for dpId=7
02:09:20.953 DMP: 55 AA 03 07 00 0C 08 00 00 08 08 79 00 00 00 00 00 00 A6
02:09:20.958 MQT: tele/tasmota_41AF7F/RESULT = {"TuyaReceived":{"Data":"55AA0307000C080000080879000000000000A6","Cmnd":7,"CmndData":"0800000808790000 00000000","DpType0Id8":"0x0879000000000000","8":{"DpId":8,"DpIdType":0,"DpIdData":"0879000000000000"}}}

we have

087A 00024A 000040 217.0V  0.586A  64W
0878 000123 00001F 216.8V  0.291A  31W
0879 000000 000000 216.9V  0.657A  49W

Now all you need is for Tasmota to decode this data and display it. It is also possible to calculate other data.

**Reply by RaW79** [reply]
> Now all you need is for Tasmota to decode this data and display it. It is also possible to calculate other data.

Yes, the philosophy behind is as shown above clear and this sounds simple. But without coding/skripting knowledge it´s a huge problem for me and unsolveable for me without hints and guidance what to look for and do. Therefore I´m asking but sadly this discussion stopped so far.
